File: kndvi.r
#if (!require("raster")) install.packages("raster")
require(raster)

# Function to convert Sentinel-2 L2A into reflectance values
s2_surface_reflectance <- function(x){
  #@x: RasterLayer, RasterStack or RasterBrick containing Sentinel-2 L2A data
  return(x/10000)
}

# Function to compute kNDVI from two RasterLayers
kndvi <- function(nir, red, sigma=0.15, method="rbf", faster=FALSE){
  #@nir: RasterLayer containing reflectance values in the Near-Infrared band
  #@red: RasterLayer containing reflectance values in the Red band
  #@sigma: sigma value. This argument accept a single sigma value or a RasterLayer containing the pixel-wise
  # sigma value. By default sigma=0.15.
  #@mode: method used to compute kNDVI. Options: "rbf" or "tanh". Default="rbf"
  #@faster: Not implemented! If faster=TRUE non-evaluate range of values and make the function faster
  if(max(raster::values(nir))>1){message("The NIR band contains values greater than 1")}
  if(min(raster::values(nir))<0){message("The NIR band contains values lower than 0")}
  if(max(raster::values(red))>1){message("The Red band contains values greater than 1")}
  if(min(raster::values(red))<1){message("The Red band contains values lower than 0")}
  if(method=="rbf"){
    knr <- exp(-(nir-red)^2/(2*sigma^2))
    kndvi <- (1-knr) / (1+knr)
  }
  if(method=="tanh"){
    kndvi <- tanh(((nir-red)/(2*sigma))^2)
  }
  return(kndvi)
}

# Function to compute the pixel-wise sigma value
pixel_wise_sigma <- function(nir, red){
  #@nir: RasterLayer containing reflectance values in the Near-Infrared band
  #@red: RasterLayer containing reflectance values in the Red band
  sigma <- 0.5 * (nir + red)
  return(sigma)
}

# Function to compute the median pixel-wise sigma value of the scene
median_dist_sigma <- function(pixel_wise_sigma, nsamples=1000){
  #@nir: RasterLayer containing reflectance values in the Near-Infrared band
  #@red: RasterLayer containing reflectance values in the Red band
  #@nsamples: Not implemented! Number of random samples used to compute the median value
  set.seed(1234)
  return(median(raster::values(pixel_wise_sigma)))
}

File: python/kndvi.py
# kNDVI with RBF kernel.
# nir: near infrared (pixel) value
# red: red value

import numpy as np
from scipy.spatial.distance import pdist


def median_dist_sigma(X: np.ndarray, subsample: int=None, seed: int=123) -> float:
    """standard median distance estimation for sigma
    Args: 
        X (np.ndarray): input data, shape=(n_samples, n_features)
        subsample (int): subsample of the data, optional
            recommended value of 1_000 for very large datasets
        seed (int): for reproducibility on the random subsample
    Returns:
        sigma (float): sigma value
    """

    # ensure 1D for pdist function
    if X.ndim == 1:
        X = np.atleast_2d(X).T
    
    # only subsample if asked for or less than number of samples
    if subsample is not None and subsample < X.shape[0]:
        # random seed for reproducibility
        rng = np.random.RandomState(seed)

        # permute samples and take a subset
        X = rng.permutation(X)[:subsample]

    # get the median distance
    sigma = np.median(pdist(X, "euclidean"))

    return sigma


def pixel_wise_sigma(nir: np.ndarray, red: np.ndarray) -> float:
    """Calculate sigma pixel-wise
    Estimate sigma independently for each pixel as sigma=0.5*(NIR+red).
    
    Args:
        nir (np.ndarray): near infrared band, shape=(n_samples)
        red (np.ndarray): red band, shape=(n_samples)
    Returns:
        sigma (float): sigma value
        """
    sigma = 0.5 * (nir + red)

    return sigma


def calculate_kndvi(nir: np.ndarray, red: np.ndarray, sigma: float=0.15) -> np.ndarray:
    """calculate kernel NDVI method
    
    Args:
        nir (np.ndarray): near infrared band, shape=(n_samples)
        red (np.ndarray): red band, shape=(n_samples)
        sigma (float): length scale for the kernel method
    Returns:
        kndvi (np.ndarray): kernel ndvi, shape=(n_samples)
    """
    # calculate knr
    knr = np.exp(-((nir - red) ** 2) / (2 * sigma ** 2))

    # calculate kndvi

    kndvi = (1 - knr) / (1 + knr)
    # An equivalent form to compute it is
    # kndvi = np.tanh(((nir-red)/(2*sigma))**2)

    return kndvi


def main():

    # estimate sigma
    n_samples = 10
    nir = np.random.randn(n_samples)
    red = 0.1 * np.random.randn(n_samples)

    sigma = pixel_wise_sigma(nir, red)

    # estimate kernel ndvi
    k_indice = calculate_kndvi(nir, red, sigma)


if __name__ == "__main__":
    main()
